packages/@react-spectrum/s2/style/spectrum-theme.ts

```diff
@@ -74,41 +74,138 @@ const color = {
   LinkText: 'LinkText'
 };
 
-export function baseColor(base: keyof typeof color): {default: keyof typeof color, isHovered: keyof typeof color, isFocusVisible: keyof typeof color, isPressed: keyof typeof color} {
-  let keys = Object.keys(color) as (keyof typeof color)[];
-  let index = keys.indexOf(base);
+// Resolves a color to its most basic form, following all aliases.
+function resolveColorToken(token: string | ColorToken | ColorRef): ColorToken {
+  if (typeof token === 'string') {
+    return {
+      type: 'color',
+      light: token,
+      dark: token
+    };
+  }
+
+  if (token.type === 'color') {
+    return token;
+  }
+
+  let lightToken = baseColors[token.light];
+  if (!lightToken) {
+    throw new Error(`${token.light} is not a valid color reference`);
+  }
+  let darkToken = baseColors[token.dark];
+  if (!darkToken) {
+    throw new Error(`${token.dark} is not a valid color reference`);
+  }
+
+  let light = resolveColorToken(lightToken);
+  let dark = resolveColorToken(darkToken);
+  return {
+    type: 'color',
+    light: light.light,
+    dark: dark.dark
+  };
+}
+
+function colorTokenToString(token: ColorToken, opacity?: string | number) {
+  if (token.light === token.dark) {
+    return token.light;
+  }
+  let result = `light-dark(${token.light}, ${token.dark})`;
+  if (opacity) {
+    result = `rgb(from ${result} r g b / ${opacity}%)`;
+  }
+  return result;
+}
+
+// Bumps up a color token by one stop, e.g. for hover/press states.
+let colorList = Object.keys(baseColors);
+function nextColorStop(name: string, token: string | ColorToken | ColorRef): ColorToken {
+  if (typeof token === 'object' && token.type === 'ref') {
+    let light = nextColorStop(token.light, baseColors[token.light]);
+    let dark = nextColorStop(token.dark, baseColors[token.dark]);
+
+    return {
+      type: 'color',
+      light: light.light,
+      dark: dark.dark,
+      forcedColors: token.forcedColors
+    };
+  }
+
+  let index = colorList.indexOf(name);
   if (index === -1) {
-    throw new Error('Invalid base color ' + base);
+    throw new Error(`${name} does not support states`);
+  }
+
+  let key = colorList[index + 1];
+  if (key.split('-')[0] !== name.split('-')[0]) {
+    throw new Error(`${name} does not support states`);
```

Comment:
difference between this error and the previous one? is the first one actually that the color doesn't exist and this one is that there isn't a state?

Reply:
e.g. if you have `gray-1000`, you can't apply a hover state to that because `gray-1100` does not exist. The next color after it would be in a different scale (e.g. `blue-100` or something). So this checks if they are in the same scale.